File: src/Clap Detection/ClapAll/first.py
# #!/usr/bin/env python3
# # -*- coding: utf-8 -*-
# """
# Created on Sun Apr  9 18:34:46 2023

# @author: savan
# """

# import wave
# import struct

# MIN_MATCH_LEN = 2000
# THRESHOLD = 0.7

# def find_wav_sample(wav_file, sample_file, window_size, slide_interval):
#     # Load WAV files
#     with wave.open(wav_file, 'rb') as wav_fs, wave.open(sample_file, 'rb') as sample_fs:
#         # Read WAV headers
#         print(wav_fs.getparams())
#         wav_chunk_size, wav_fmt_chunk_size, sample_rate, byte_rate, wav_data_size, wav_fmt, wav_num_channels, wav_bits_per_sample = wav_fs.getparams()
#         sample_chunk_size, sample_fmt, sample_num_channels, sample_sample_rate, sample_byte_rate, sample_bits_per_sample = sample_fs.getparams()

#         if sample_fmt != 1 or sample_num_channels != 1 or sample_bits_per_sample != 16:
#             print("Error: Sample format is not supported")
#             return []

#         # Read sample data
#         sample_data = struct.unpack('<' + str(sample_fs.getnframes()) + 'h', sample_fs.readframes(sample_fs.getnframes()))

#         # Define variables
#         buf = bytearray(2 * MIN_MATCH_LEN)
#         window_size_samples = int(window_size * sample_sample_rate)
#         slide_interval_samples = int(slide_interval * sample_sample_rate)
#         threshold = int(THRESHOLD * 32767)

#         # Iterate over the WAV file
#         matches = []
#         sample_buf = bytearray(2 * window_size_samples)
#         wav_fs.setpos(44)
#         i = 0
#         while wav_fs.readframes(MIN_MATCH_LEN) != b'':
#             wav_fs.readinto(buf)
#             read_size = len(buf) // 2
#             j = 0
#             while j + window_size_samples <= read_size:
#                 # Extract the current window of samples from the WAV file
#                 window_buf = buf[j*2:(j+window_size_samples)*2]
#                 # Extract the corresponding window of samples from the sample data
#                 sample_start = i * slide_interval_samples
#                 sample_buf = sample_data[sample_start:sample_start+window_size_samples]

#                 # Compute the correlation between the two windows of samples
#                 corr = 0
#                 for k in range(window_size_samples):
#                     corr += sample_buf[k] * window_buf[k*2] + sample_buf[k] * window_buf[k*2+1]
#                 corr = abs(corr) / (window_size_samples * 32767)

#                 # Check if the correlation is above the threshold
#                 if corr >= threshold:
#                     matches.append(i * slide_interval_samples)
#                 j += slide_interval_samples // 2
#             i += 1

#     return matches


# matches = find_wav_sample('original.wav', 'original.wav', 0.1, 0.01)
# if len(matches) >= 7:
#     print("Match found!")
# else:
#     print("No match found.")


# # with wave.open('original.wav', 'rb') as wav_file:
# #     print('Number of channels:', wav_file.getnchannels())
# #     print('Sample rate:', wav_file.getframerate())
# #     print('Sample width (in bytes):', wav_file.getsampwidth())
# #     print('Encoding:', wav_file.getcomptype())


# # with wave.open('sample.wav', 'rb') as wav_file:
# #     print('Number of channels:', wav_file.getnchannels())
# #     print('Sample rate:', wav_file.getframerate())
# #     print('Sample width (in bytes):', wav_file.getsampwidth())
# #     print('Encoding:', wav_file.getcomptype())









# import numpy as np
# import soundfile as sf

# def find_wav_sample(original_file, sample_file, window_size, slide_interval):
#     # Load the original file and sample file
#     original, sr_original = sf.read(original_file)
#     sample, sr_sample = sf.read(sample_file)

#     # Compute the number of samples that the sliding window should cover
#     window_samples = int(window_size * sr_sample)

#     # Compute the number of samples to slide the window by
#     slide_samples = int(slide_interval * sr_sample)

#     # Compute the maximum number of windows to slide
#     max_slides = len(sample) - window_samples

#     # Compute the similarity score for each window and find the index of the maximum score
#     max_score = -1
#     max_score_index = -1
#     for i in range(0, max_slides, slide_samples):
#         window = sample[i:i+window_samples]
#         score = np.sum(np.abs(original - window)) / np.sum(np.abs(original))
#         if score > max_score:
#             max_score = score
#             max_score_index = i

#     # Compute the percentage match
#     num_claps_original = 10
#     num_claps_sample = int(max_score * num_claps_original)
#     percentage_match = num_claps_sample / num_claps_original

#     return percentage_match

# # Example usage
# original_file = 'original.wav'
# sample_file = 'sample.wav'
# window_size = 0.1 # 100ms
# slide_interval = 0.01 # 10ms

# percentage_match = find_wav_sample(original_file, sample_file, window_size, slide_interval)
# if percentage_match >= 0.7:
#     print('Sample matches with original.')
# else:
#     print('Sample does not match with original.')



import math
import wave

def select_window_and_slide(original_file, sample_file, window_size_min=0.01, slide_interval_min=0.001):
    # Get the length of the audio files
    original_length = get_audio_length(original_file)
    sample_length = get_audio_length(sample_file)

    # Compute the maximum window size and slide interval based on the audio lengths
    window_size_max = min(original_length, sample_length)
    slide_interval_max = min(window_size_max, 1.0)

    # Compute the number of windows and slides required for each file
    num_windows_original = math.ceil((original_length - window_size_min) / slide_interval_min)
    num_windows_sample = math.ceil((sample_length - window_size_min) / slide_interval_min)

    # Compute the actual window size and slide interval for each file
    window_size_original = (original_length - window_size_min) / (num_windows_original - 1)
    slide_interval_original = (original_length - window_size_original) / (num_windows_original - 1)
    window_size_sample = (sample_length - window_size_min) / (num_windows_sample - 1)
    slide_interval_sample = (sample_length - window_size_sample) / (num_windows_sample - 1)

    # Choose the smaller of the two window sizes and the larger of the two slide intervals
    window_size = min(window_size_original, window_size_sample)
    slide_interval = max(slide_interval_original, slide_interval_sample)

    return window_size, slide_interval

def get_audio_length(audio_file):
    with wave.open(audio_file, 'r') as f:
        frames = f.getnframes()
        rate = f.getframerate()
        return frames / float(rate)



import numpy as np
import wave

def find_wav_sample(original_wav_file, sample_wav_file, threshold=0.1):
    original_audio = load_wav(original_wav_file)
    sample_audio = load_wav(sample_wav_file)

    # Set the window size and slide interval
    window_size = int(len(original_audio) * 0.1)
    slide_interval = int(window_size * 0.5)

    # Compute the dot product between the original audio and the sample audio
    dot_products = []
    for i in range(0, len(original_audio) - window_size, slide_interval):
        window = original_audio[i:i+window_size]
        # Pad the sample audio if the window size is longer than the sample audio
        if len(window) > len(sample_audio):
            padded_sample_audio = np.concatenate((sample_audio, np.zeros(len(window) - len(sample_audio))))
            dot_product = np.dot(window, padded_sample_audio)
        else:
            dot_product = np.dot(window, sample_audio[:len(window)])
        dot_products.append(dot_product)

    # Compute the maximum dot product and the threshold value
    max_dot_product = max(dot_products)
    threshold_value = threshold * max_dot_product

    # Check if the maximum dot product exceeds the threshold
    if max_dot_product >= threshold_value:
        return True
    else:
        return False




def load_wav(audio_file):
    # Load the audio file
    with wave.open(audio_file, 'rb') as f:
        num_frames = f.getnframes()
        channels = f.getnchannels()
        sample_width = f.getsampwidth()
        frame_rate = f.getframerate()
        frames = f.readframes(num_frames)

    # Convert the raw audio bytes to a numpy array
    if sample_width == 1:
        dtype = np.uint8
    else:
        dtype = np.int16
    audio = np.frombuffer(frames, dtype=dtype)

    print('audio shape:', audio.shape)
    print('audio contents:', audio)

    # Reshape the audio array into a two-dimensional array with one column for each channel
    audio = np.reshape(audio, (num_frames, channels))

    # Convert the audio array to mono
    audio = np.mean(audio, axis=1)

    return audio


window_size, slide_interval = select_window_and_slide('original.wav', 'sample.wav')
# if find_wav_sample('original.wav', 'sample.wav'):
#     print("Match found!")
# else:
#     print("No match found.")












import numpy as np
import scipy.signal as signal
import soundfile as sf

# Load the original audio file
audio_orig, sr_orig = sf.read('original.wav')

# Load the sample audio file
audio_sample, sr_sample = sf.read('sample.wav')

# Define the filter parameters
order = 1000
f_start = 0.0
f_end = sr_sample / 2
num_freqs = order // 2 + 1

# Create the frequency vector
freq = np.linspace(f_start, f_end, num_freqs, endpoint=False)

# Define the filter
filter_coeffs = signal.firwin(order, freq, pass_zero=False)

# Apply the filter to the original audio
audio_filtered = signal.convolve(audio_orig, filter_coeffs, mode='same')

# Compute the correlation between the filtered audio and the sample audio
corr = signal.correlate(audio_filtered, audio_sample)

# Check if the correlation exceeds a certain threshold
threshold = 0.7
if np.max(corr) > threshold:
    print('Match found!')
else:
    print('Match not found.')

File: src/Clap Detection/ClapAll/first.cpp
// // #include <iostream>
// // #include <vector>
// // #include <cmath>
// // #include <fstream>
// // #include <algorithm>


// // // Define a function to calculate the percentage of match between two audio files
// // double find_per(const std::string& within_file, const std::string& find_file, int window) {
// //     std::ifstream within_fs(within_file, std::ios::binary);
// //     std::ifstream find_fs(find_file, std::ios::binary);

// //     // Get the sample rate of the within_file
// //     uint32_t sr_within;
// //     within_fs.seekg(24);
// //     within_fs.read(reinterpret_cast<char*>(&sr_within), sizeof(sr_within));

// //     // Read the audio data from the within_file and the find_file
// //     std::vector<int16_t> y_within(sr_within * window);
// //     std::vector<int16_t> y_find(sr_within * window);
// //     within_fs.seekg(44);
// //     within_fs.read(reinterpret_cast<char*>(&y_within[0]), y_within.size() * sizeof(y_within[0]));
// //     find_fs.seekg(44);
// //     find_fs.read(reinterpret_cast<char*>(&y_find[0]), y_find.size() * sizeof(y_find[0]));

// //     // Calculate the correlation between the two audio signals
// //     std::vector<double> c(y_within.size() + y_find.size() - 1);
// //     std::cout<< "Hey\n";
// //     std::cout << "The size of the vector is: " << c.size() << std::endl;
// //     for (int i = 0; i < c.size(); i++) {
// //         c[i] = 0;
// //         for (int j = 0; j < y_within.size(); j++) {
// //             if (i - j >= 0 && i - j < y_find.size()) {
// //                 c[i] += y_within[j] * y_find[i - j];
// //             }
// //         }
// //         c[i] = std::abs(c[i]);
// //     }
// //     std::cout << "The size of the vector is: " << c.size() << std::endl;
// //     std::cout<< "Hey\n";
// //     // Square the correlation values
// //     for (int i = 0; i < c.size(); i++) {
// //         c[i] *= c[i];
// //     }

// //     // Find the threshold and count the number of matches
// //     double thres = 0.25 * (*std::max_element(c.begin(), c.end()));
// //     int count = 0;
// //     for (int i = 0; i < c.size(); i++) {
// //         if (c[i] > thres) {
// //             count++;
// //         }
// //     }

// //     // Calculate the percentage of matches
// //     return static_cast<double>(count) / c.size() * 100.0;
// // }

// // // Define the main function
// // // int main(int argc, char** argv) {
// // //     // Parse command line arguments
// // //     int window = 10;
// // //     std::string within_file = "sample.wav";
// // //     std::string find_file = "original.wav";
// // //     if (argc > 1) {
// // //         window = std::stoi(argv[1]);
// // //     }
// // //     if (argc > 2) {
// // //         within_file = argv[2];
// // //     }
// // //     if (argc > 3) {
// // //         find_file = argv[3];
// // //     }

// // //     // Calculate the percentage of matches and print the result
// // //     double per = find_per(within_file, find_file, window);
// // //     if (per <= 0.5) {
// // //         std::cout << "Match Found\n";
// // //     } else {
// // //         std::cout << "Match not found\n";
// // //     }

// // //     return 0;
// // // }

// // int main() {
// //     // Parse command line arguments
// //     int window = 10;
// //     std::string within_file = "sample.wav";
// //     std::string find_file = "original.wav";

// //     // Calculate the percentage of matches and print the result
// //     std::cout<< "Hey\n";
// //     double per = find_per(within_file, find_file, window);
// //     std::cout<< "Hey\n";
// //     if (per <= 0.5) {
// //         std::cout << "Match Found\n";
// //     } else {
// //         std::cout << "Match not found\n";
// //     }

// //     return 0;
// // }







// #include <iostream>
// #include <vector>
// #include <cmath>
// #include <fstream>
// #include <algorithm>

// // Define a function to calculate the percentage of match between two audio files
// double find_per(const std::string& within_file, const std::string& find_file, int window) {
//     std::ifstream within_fs(within_file, std::ios::binary);
//     std::ifstream find_fs(find_file, std::ios::binary);

//     // Get the sample rate of the within_file
//     uint32_t sr_within;
//     within_fs.seekg(24);
//     within_fs.read(reinterpret_cast<char*>(&sr_within), sizeof(sr_within));

//     // Read the audio data from the within_file and the find_file
//     std::vector<int16_t> y_within(sr_within * window);
//     std::vector<int16_t> y_find(sr_within * window);
//     within_fs.seekg(44);
//     within_fs.read(reinterpret_cast<char*>(&y_within[0]), y_within.size() * sizeof(y_within[0]));
//     find_fs.seekg(44);
//     find_fs.read(reinterpret_cast<char*>(&y_find[0]), y_find.size() * sizeof(y_find[0]));

//     // Calculate the correlation between the two audio signals
//     std::vector<double> c(y_within.size() + y_find.size() - 1);
//     std::cout << "The size of the vector is: " << c.size() << std::endl;
//     for (int i = 0; i < y_within.size(); i++) {
//         for (int j = 0; j < y_find.size(); j++) {
//             c[i + j] += y_within[i] * y_find[j];
//         }
//     }
//     std::cout << "The size of the vector is: " << c.size() << std::endl;
//     // Square the correlation values
//     std::for_each(c.begin(), c.end(), [](double& x){ x *= x; });

//     // Find the threshold and count the number of matches
//     double thres = 0.25 * (*std::max_element(c.begin(), c.end()));
//     int count = std::count_if(c.begin(), c.end(), [thres](double x){ return x > thres; });

//     // Calculate the percentage of matches
//     return static_cast<double>(count) / c.size() * 100.0;
// }

// int main() {
//     // Parse command line arguments
//     int window = 10;
//     std::string within_file = "sample.wav";
//     std::string find_file = "original.wav";

//     // Calculate the percentage of matches and print the result
//     std::cout<< "Hey\n";
//     double per = find_per(within_file, find_file, window);
//     std::cout<< "Hey\n";
//     if (per <= 0.5) {
//         std::cout << "Match Found\n";
//     } else {
//         std::cout << "Match not found\n";
//     }

//     return 0;
// }






#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>
#include <algorithm>
#include <cstdint> // for uint32_t

// Define a function to calculate the percentage of match between two audio files
double find_per(const std::string& within_file, const std::string& find_file, int window) {
    std::ifstream within_fs(within_file, std::ios::binary);
    std::ifstream find_fs(find_file, std::ios::binary);

    // Get the sample rate of the within_file
    uint32_t sr_within;
    within_fs.seekg(24);
    within_fs.read(reinterpret_cast<char*>(&sr_within), sizeof(sr_within));

    // Read the audio data from the within_file and the find_file
    std::vector<int16_t> y_within(sr_within * window);
    std::vector<int16_t> y_find(sr_within * window);
    within_fs.seekg(44);
    within_fs.read(reinterpret_cast<char*>(&y_within[0]), y_within.size() * sizeof(y_within[0]));
    find_fs.seekg(44);
    find_fs.read(reinterpret_cast<char*>(&y_find[0]), y_find.size() * sizeof(y_find[0]));

    // Calculate the correlation between the two audio signals
    std::vector<double> c(y_within.size() + y_find.size() - 1);
    std::cout<< "Hey\n";
    std::cout << "The size of the vector is: " << c.size() << std::endl;
    // for (int i = 0; i < c.size(); i++) {
    //     c[i] = 0;
    //     for (int j = 0; j < y_within.size(); j++) {
    //         if (i - j >= 0 && i - j < y_find.size()) {
    //             c[i] += y_within[j] * y_find[i - j];
    //         }
    //     }
    //     c[i] = std::abs(c[i]);
    // }

    for (int i = 0; i < c.size(); i++) {
        c[i] = 0;
        int j_start = std::max(0, i - (int)y_within.size() + 1);
        int j_end = std::min((int)y_find.size() - 1, i);
        for (int j = j_start; j <= j_end; j++) {
            c[i] += y_within[i - j] * y_find[j];
        }
        c[i] = std::abs(c[i]);
    }

    std::cout << "The size of the vector is: " << c.size() << std::endl;
    std::cout<< "Hey\n";
    // Square the correlation values
    for (int i = 0; i < c.size(); i++) {
        c[i] *= c[i];
    }

    // Find the threshold and count the number of matches
    double thres = 0.25 * (*std::max_element(c.begin(), c.end()));
    int count = 0;
    for (int i = 0; i < c.size(); i++) {
        if (c[i] > thres) {
            count++;
        }
    }

    // Calculate the percentage of matches
    return static_cast<double>(count) / c.size() * 100.0;
}

// Define the main function
int main(int argc, char** argv) {
    // Parse command line arguments
    int window = 10;
    std::string within_file = "sample.wav";
    std::string find_file = "original.wav";
    if (argc > 1) {
        window = std::stoi(argv[1]);
    }
    if (argc > 2) {
        within_file = argv[2];
    }
    if (argc > 3) {
        find_file = argv[3];
    }

    // Calculate the percentage of matches and print the result
    double per = find_per(within_file, find_file, window);
    if (per <= 0.5) {
        std::cout << "Match Found\n";
    } else {
        std::cout << "Match not found\n";
    }

    return 0;
}